src/main/java/bio/terra/stairway/FlightDao.java

```diff
@@ -55,144 +56,43 @@
-
-      commitTransaction(connection);
-      return stairwayId;
-    } catch (SQLException ex) {
-      throw new DatabaseOperationException("Stairway instance lookup failed", ex);
-    }
-  }
-
-  private String lookupStairwayInstanceQuery(Connection connection, String stairwayName)
-      throws SQLException, DatabaseOperationException {
-    final String sqlStairwayInstance =
-        "SELECT stairway_id"
-            + " FROM "
-            + STAIRWAY_INSTANCE_TABLE
-            + " WHERE stairway_name = :stairwayName";
-
-    try (NamedParameterPreparedStatement instanceStatement =
-        new NamedParameterPreparedStatement(connection, sqlStairwayInstance)) {
-
-      instanceStatement.setString("stairwayName", stairwayName);
-
-      try (ResultSet rs = instanceStatement.getPreparedStatement().executeQuery()) {
-        List<String> stairwayIdList = new ArrayList<>();
-        while (rs.next()) {
-          stairwayIdList.add(rs.getString("stairway_id"));
-        }
-
-        if (stairwayIdList.size() == 0) {
-          return null;
-        }
-        if (stairwayIdList.size() > 1) {
-          throw new DatabaseOperationException("Multiple stairways with the same name!");
-        }
-        return stairwayIdList.get(0);
-      }
-    }
-  }
-
-  /**
-   * This call can be used in conjunction with a list of the currently active stairway instances to
-   * drive recovery of orphaned flights.
+   * Record a new flight
    *
-   * @return list of the stairway instances known to stairway
-   * @throws DatabaseOperationException on SQL exception
+   * @param flightContext description of the flight
+   * @throws DatabaseOperationException database error
+   * @throws InterruptedException thread shutdown
    */
-  List<String> getStairwayInstanceList() throws DatabaseOperationException {
-    final String sql = "SELECT stairway_name FROM " + STAIRWAY_INSTANCE_TABLE;
-    List<String> instanceList = new ArrayList<>();
-
-    try (Connection connection = dataSource.getConnection()) {
-      startReadOnlyTransaction(connection);
-
-      try (NamedParameterPreparedStatement instanceStatement =
-              new NamedParameterPreparedStatement(connection, sql);
-          ResultSet rs = instanceStatement.getPreparedStatement().executeQuery()) {
-        while (rs.next()) {
-          instanceList.add(rs.getString("stairway_name"));
-        }
-      }
-
-      commitTransaction(connection);
-    } catch (SQLException ex) {
-      throw new DatabaseOperationException("Stairway instance enumeration failed", ex);
-    }
-    return instanceList;
-  }
-
-  /** Record a new flight */
   void submit(FlightContext flightContext)
       throws DatabaseOperationException, InterruptedException, DuplicateFlightIdSubmittedException {
+    DbRetry dbRetry = new DbRetry("flight.submit");
+    dbRetry.perform(() -> submitInner(flightContext));
+  }
+
+  private Void submitInner(FlightContext flightContext)
```

Comment:
I'm not sure how well this would work, but you could have two `perform` methods, with one that supports `void` methods. That would avoid the need for `Void` and `return null;`

Reply:
From what I read, an `@FunctionalInterface` has to have a return value. The recommendation was to return Void, but you cannot `new` a Void, so you return null.

Is there a variant of functional interface that does not return a value?

Reply:
Not sure where this info is in the java spec, but `void` must be allowed as `Consumer` is a functional interface.

Reply:
FYI here's the relevant part of the JLS https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.8

Reply:
Got it. It is a little bit of repetition in the DbRetry class, but gets rid of the ugly `return null` in the main code. Seems like a good trade-off.